src/Microsoft.XmlSerializer.Generator/src/SgenTasks.cs

```diff
@@ -0,0 +1,90 @@
+﻿using System;
+using System.IO;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.Build.Utilities;
+using Microsoft.Build.Framework;
+
+namespace Microsoft.XmlSerializer.Generator
+{
+    public class SgenParameters : Task
+    {
+        public string Types { get; set; }
+        public string References { get; set; }
+        public bool UseProxyTypes { get; set; }
+        public bool Verbose { get; set; }
+        public string KeyFile { get; set; }
+        public string KeyContainer { get; set; }
+        public bool DelaySign { get; set; }
+
+        public override bool Execute()
+        {
+            var sgenCommands = new List<string>();
+            if(Types != null)
+                sgenCommands.Add("--type " + Types);
+            if (UseProxyTypes)
+                sgenCommands.Add("--proxytypes");
+            if (Verbose)
+                sgenCommands.Add("--verbose");
+            if (References != null)
+            {
+                sgenCommands.Add("--reference " + References);
+            }
+
+            var cscCommands = new List<string>();
+            if (KeyFile != null)
+                cscCommands.Add("-keyfile:" + KeyFile);
+            if (KeyContainer != null)
+                cscCommands.Add("-keycontainer:" + KeyContainer);
+            if (DelaySign)
+                cscCommands.Add("-delaysign");
+
+            try
+            {
+                if(sgenCommands.Count > 0)
+                {
+                    string sgenRspPath = Environment.GetEnvironmentVariable("SgenRspFilePath");
+                    File.WriteAllLines(sgenRspPath, sgenCommands.ToArray());
+                }
+
+                if(cscCommands.Count > 0)
+                {
+                    string cscRspPath = Environment.GetEnvironmentVariable("SgenCscRspFilePath");
+                    File.WriteAllLines(cscRspPath, cscCommands.ToArray());
+                }
+            }
+            catch (Exception e)
+            {
+                //sgenparameter task won't break the build even when exceptions are thrown, it will continue create serializer using default settings
+                Log.LogError(string.Format("Unable to generate temporary response file, error: {0}", e.Message));
+                Log.LogError("SgenParameters task failed. Serializer is generating using default parameters.");
+                return true;
+            }
+           
+            return true;
+        }
+        
+    }
+
+    public class SgenSetEnv : Task
```

Comment:
Why are you communicating via environment variables?

Reply:
I see, you're relying on these to be set before some users project calls SgenParameters.

This behavior is very fragile and unusual.  Why didn't you just document some properties that folks set in order to control the behavior of your GenerateSerializationAssembly target?  That's how nearly every other MSBuild-based target provides for extensibility.

Reply:
@ericst Because in order to prevent name conflicts, those properties need to have kind of long names, I think it doesn't look very pretty... That's the only reason. So I'm trying to find a way to make it looks neat from customer's perspective.

Reply:
An `SGen` prefix doesn't seem too ugly from my perspective.  As an alternative you could have them define metadata on an ItemDefinition group for SGen items.  This is similar to what the VCXProj system does:
```xml
<ItemDefinitionGroup>
  <SGenParameters>
      <Reference>%(Reference);myreferencepath.dll</Reference>
      <Defines>%(Defines);FEATURE_FOO</Defines>
  </SGenParameters>
</ItemDefinitionGroup>
  ````

Or even better, if you get a new MSBuild with https://github.com/Microsoft/msbuild/issues/3833 you could do:
```xml
<ItemDefinitionGroup>
  <SGenParameters
    Reference="%(Reference);myreferencepath.dll"
    Defines="%(Defines);FEATURE_FOO" />
</ItemDefinitionGroup>

```
Though that might feel a little *off* for C# developers who likely have never seen an ItemDefinitionGroup before.  @rainersigwald might have some other suggestions.

Honestly, properties are good enough for the Compiler and all the other built-in tools in common.targets, why aren't good enough for SGen?  Having to run a target to run a task to write a temporary response file might look pretty from the sample perspective but its definitely not intuitive.  Here are a handful of other problems to think about:
1. Where does the target sequence itself and will it always run in the right place?
2. What if I only want to *augment* one parameter and not replace the all parameters?
3. What if more than one entity wants to contribute to these parameters (eg: a package and the project itself)?
4. What happens when a child process inherits the environment variables you set?
5.  I could probably go on, but suffice it to say you're doing something very "special" here and it's likely got a lot of side-effects that you'd want to think about.